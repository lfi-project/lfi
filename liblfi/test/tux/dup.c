#include <unistd.h>
#include <stdio.h>
#include <fcntl.h>
#include <string.h>

int main() {
    // Test dup() syscall
    int fd1 = open("/dev/null", O_WRONLY);
    if (fd1 < 0) {
        printf("failed to open /dev/null\n");
        return 1;
    }
    
    int fd2 = dup(fd1);
    if (fd2 < 0) {
        printf("dup() failed\n");
        return 1;
    }
    
    printf("dup: original fd=%d, dup fd=%d\n", fd1, fd2);
    
    // Test dup3() syscall
    int fd3 = 10; // arbitrary target fd
    int fd4 = dup3(fd1, fd3, 0);
    if (fd4 != fd3) {
        printf("dup3() failed: expected %d, got %d\n", fd3, fd4);
        return 1;
    }
    
    printf("dup3: original fd=%d, target fd=%d\n", fd1, fd3);
    
    // Test writing to all fds to verify they work
    const char* msg = "test\n";
    write(fd1, msg, strlen(msg));
    write(fd2, msg, strlen(msg));
    write(fd3, msg, strlen(msg));
    
    printf("dup test passed\n");
    return 0;
}