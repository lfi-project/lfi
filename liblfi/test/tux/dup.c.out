dup: original fd=3, dup fd=4
dup3: original fd=3, target fd=10
dup test passed
